### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by sjappleton** [De-Man/silvia-pi#1]
Will you be adding the multi threading still to this code? I am trying to build something similar but have not been able to get it to multi thread correctly with Kivy.

**Comment by De-Man** [comment]
Hi sjappleton, I haven’t looked at this code for awhile. 

The build I am currently running on my Silvia-Pi with Kivy is multithreaded. I’m currently in the middle of no where so I’ll take a look when I can get on my comp.

However, if I was to do this again I wouldn’t use Kivy, I’d use a node.js setup or something, just makes building the GUI quicker with less code. Kivy I found to be  temperamental...

**Comment by sjappleton** [comment]
Hi Thanks for the response. 

I would love to see that when possible. No worries about the time I have been messing with my Gaggia for several months so I am in no hurry. I understand the point on Kivy it is quite difficult to play with. I like the result as I like the touchscreen controls.

Cheers

**Comment by De-Man** [comment]
You’re welcome, I’ll check it out when I get home and let you know.

**Comment by sjappleton** [comment]
Any chance you have been able to look at that multi threading?

**Comment by De-Man** [comment]
hey mate, sorry about the late reply. I'll have a look now.

**Comment by De-Man** [comment]
Hey Sjappleton, unfortunately my HDs that I kept all my old stuff on is back in Australia. And I'm not going to be back for a couple of months. In these tuts I didn't include the multithreaded part and I always intended on doing another tutorial on it.

But to be honest threading is super easy to implement in python, I can help you if you like:

follow this:

1. Tutorial on threading: [here](https://realpython.com/intro-to-python-threading/)
2. You will need 2 threads. 1 for temperature in, and 1 for everything else.

Out of interest, where are you currently at on your project? what is the complete intent/plan?

**Comment by sjappleton** [comment]
I would appreciate that. I have a Gaggia Classic Espresso machine that I am modifying. End goal is a touch screen PID controlled espresso machine with shot timer. I have the basic Kivy screen designed with the shot timer and relays working but I am having difficulty incorporating the PID control/temperature without things really slowing down. I will then be mounting the raspberry pi touchscreen unit to my espresso machine with a teak wood facade and custom wood handles for the steam and portafilter.

I will post what I have. I am a mechanical engineer with background in controls and the electrical side but my coding skills are pretty amateur at best.

**Comment by De-Man** [comment]
Perfect. Yeah post what you have including photos, I am interested to see what it looks like so far.

**Couple of questions:**
- Are you using RPi and the RPi endorsed touch screen?
- Are you you using SSR for your temp control?
- You said your touch screen and relays are hooked up, does this mean when pressing a button in Kivy your relays are switching?
- Are you using the graph library in kivy?
- FYI: even though we are trying to make a "PID" controller what you will find is it will actually be a kind of really basic fuzzy logic controller. With more P and D and no I.... when you start testing it you'll see what I mean, this is just because of the thermal and general system lag  (I assume your set up will be similar to mine).

**Comment by sjappleton** [comment]
Yes using the RPi and 7” RPi touchscreen.

SSR relay for the heater and pump which are actually switching from my kivy program.

Not sure, I do not have it in there now but might want to add it to a separate screen.

**Comment by De-Man** [comment]
Real time plotting helps tuning the controller much easier, so something worth thinking about. It also looks cool on you touch screen.

I only used a single SSR for the heating element as the controller can switch it on an off beyond a typical solenoid relay's design capability. So I question whether or not you need an SSR for the pump? 

Well let me know what specifically you want a hand with and I will try to help as much as I can.

**Comment by sjappleton** [comment]
I will get you pictures as well this weekend but wanted to show the code such as it is. 

I think I got all the components and gui working but I am having trouble getting the controller to work in the background

**Comment by sjappleton** [comment]
Test of gui and rtd temp sensor.


-------- Original message --------

**Comment by sjappleton** [comment]
Hello,

Any chance you took a look at the code? I know it is not great but I am looking to get it working and any help would be appreciated.

Thanks
Sean


-------- Original message --------
From: Sean Appleton <SAppleton@qmplastics.com>
Date: 2019-09-05 9:07 PM (GMT-06:00)
To: De-Man/silvia-pi <reply@reply.github.com>, De-Man/silvia-pi <silvia-pi@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: Re: [De-Man/silvia-pi] Multithreading (#1)

Test of gui and rtd temp sensor.


-------- Original message --------

**Comment by De-Man** [comment]
Hi sjappleton, my apologies for my tardy-ness here. I have had a quick skim through it. I haven't had a chance to run it yet.

If you can give me 2 weeks I will be back in Oz and I can just give you my code?

**Comment by sjappleton** [comment]
Hi,

No problem. I appreciate the help and I have no issues waiting.

Cheers

**Comment by De-Man** [comment]
Hi

I have sent you the link to the private repo for the code I am currently running on my Silvia.

Let me know if you have any queries.

**Comment by sjappleton** [comment]
Thanks I really appreciate it. I will try it out soon as I can.

See the pictures of the teak wood front mount for the rpi and new custom portafilter handle i have been building.

Cheers


-------- Original message --------

**Comment by De-Man** [comment]
No probs at all.

Sorry, no pictures were showing?

**Comment by sjappleton** [comment]
So I am looking at plotter25.py and it looks like I will have to adjust the temperature as I am getting it from a max31865. Besides that I am guess I can just install this onto the RPI.

Here are the images. 
![20191025_212204_resized](https://user-images.githubusercontent.com/26077609/68091598-52fdab00-fe47-11e9-8045-5b36f27260c5.jpg)
![20191030_210354_resized](https://user-
images.githubusercontent.com/26077609/68091599-52fdab00-fe47-11e9-8f47-16fc588d22c7.jpg)
![20191102_123402_resized](https://user-images.githubusercontent.com/26077609/68091600-53964180-fe47-11e9-838a-78da26c4e2dd.jpg)
![20191102_123410_resized](https://user-images.githubusercontent.com/26077609/68091601-53964180-fe47-11e9-833c-34a4f8501952.jpg)

**Comment by sjappleton** [comment]
![20191030_210354_resized](https://user-images.githubusercontent.com/26077609/68091776-99540980-fe49-11e9-985a-e2dbc2b5b863.jpg)

**Comment by sjappleton** [comment]
I have the screen and rpi mounted and am trying to implement the program but getting an error from kivy saying it cannot find root.kv? I have kivy and garden is there another prerequisite I am missing?


-------- Original message --------

**Comment by De-Man** [comment]
Hi Sjappleton,

**Comment by De-Man** [comment]
Hi Sjappleton

Dan here. This is all looking amazing. Well done. Love the pictures. I am going to re-open this thread so I can try help you out (plus this is the public repo, others can't see the plotter25 repo), I'll make it public once I sanitise it and I'll add it to this repo. So please continue to post your issue here. Feel free to move your root.kv issue in to a new issue in this repo.

For this: 

> So I am looking at plotter25.py and it looks like I will have to adjust the temperature as I am getting it from a max31865. Besides that I am guess I can just install this onto the RPI.

Can you please answer the following as it doesn't look like you are getting temperature readings:
- are you using the max31865? I think that is what you suggested above, I can also see it in your first photo?
- what temp prob you are using? If you are using the Adafruit max31865 I also assume you are using a PT100 RTD? I used a K-type probe. This is no problem the PT100 will be far better than my K-type. 
- I assume you have already followed this tutorial?
[https://learn.adafruit.com/adafruit-max31865-rtd-pt100-amplifier/pinouts](https://learn.adafruit.com/adafruit-max31865-rtd-pt100-amplifier/pinouts)
   - Have you have set the pins up correctly as per SPI on the RPi?
   - Are you getting accurate reading with the RPi with the Max and the Pt100 probe direct from the terminal? i.e. have you:
      - got OneWire working through your Pi?
      - If yes, have you: loaded the w1-gpio and w1-term into the PI?
      - if yes, have you: found out what address the Max is loading the data to the probe. For me, as you can see in my code, the directory was: 
`$/sys/bus/w1/devices` 
the device folder was for me (this I expect to be different for you), this may be a reason why you temp isn't being read and the graph is flat: 
`3b-000000191766`

Thank you

Dan

**Comment by De-Man** [comment]
![sjappleton kivy PID espresso](https://user-images.githubusercontent.com/18529630/68994224-67df3300-0878-11ea-82da-2a7d531da4e7.jpg)

> So I am looking at plotter25.py and it looks like I will have to adjust the temperature as I am getting it from a max31865. Besides that I am guess I can just install this onto the RPI.

Yes there was an additional file (root.kv), and unfortunately I'm no where near my machine to pull it off of it (poor file management on my behalf - forgive me). 

Comment out root.kv this was a tacky way to load up the background at the time. From what I can see from your other photos the  background isn't being loaded up (I have inserted your image from the other repo above so others can see - looking good!). 

To load up the background use the following:

```
	def build(self):  # From Tutorial 1
		# Tutorial 2 step 3
		root = FloatLayout()
		
		# Tutorial 2 step 1
		# this ensures on size and position changes the background changes accordingly
		root.bind(size=self._update_rect, pos=self._update_rect)	
		# this ensures on build of the app the backgound is initiated 'before' anything else
		with root.canvas.before:
			self.rect = Rectangle(size=root.size, pos=root.pos, source='Background.png')
...
```

Specifically this:

```
		with root.canvas.before:
			self.rect = Rectangle(size=root.size, pos=root.pos, source='Background.png')
```

Make sure you do it near the top of `def build(self):`.

**Comment by sjappleton** [comment]
That makes sense I will continue to post here.

I am using a pt100 with the max31865. I like it's performance and used it on a rpi controlled meat smoker I built.

I probably made an error in that. I will take a look and I added a branch for my modification using a little program that I had from someone when i made the smoker.




-------- Original message --------

**Comment by De-Man** [comment]
> That makes sense I will continue to post here. I am using a pt100 with the max31865. I like it's performance and used it on a rpi controlled meat smoker I built. I probably made an error in that. I will take a look and I added a branch for my modification using a little program that I had from someone when i made the smoker.

Yeah, so the current problems I see:
1. No temperature reading, which should be an easy fix for you as you have already used a similar setup before.
2. No background. another easy fix.

once you get 1. & 2. sorted it will be interesting to see how the multithreaded temp reading performs.

**Comment by sjappleton** [comment]
Getting something odd now and not sure what I did.  Now seeing an error that says typeError: object.__init__() takes no parameters.

Very strange as I cannot get it back to what I had before.

Any thoughts?

**Comment by De-Man** [comment]
are you able to share your code? sounds like you're trying to initiate an object with parameters?

**Comment by sjappleton** [comment]
I will as soon as I get back home from work. The only thing I have been trying to do is to get this working with the max31865.

Must of done something stupid.

**Comment by sjappleton** [comment]
I updated the for-sjappleton code with what I had

**Comment by sjappleton** [comment]
I think i had an issue with my kivy garden and python. Using python2.7 it worked again but just need to correct the display so i can see the set point temp, timer and temp etc.

Do you have a image of what your plotter25 looks like when everything is working?


-------- Original message --------

**Comment by sjappleton** [comment]
![20191119_070415](https://user-images.githubusercontent.com/26077609/69172239-170f4a80-0ac3-11ea-9222-8fb3f7bda36b.jpg)
I got the temps displaying and within .6 degrees of another temp sensor. Was responding pretty quickly and handled 20 to 100 degrees. Need some work yet but getting there.

**Comment by De-Man** [comment]
Perfect, this is looking good. You need to load the background up so you can see everything.

Worth checking the following is working:
- buttons are pressing depressing with the correct graphic. 
- on press of coffee, the timer comes up.
- Shut down and restart functionality with the power button (bottom left).
- Screen sleep (I'd check this very last - automatically puts the screen to sleep after ## mins).

Good to hear the temp lag is good. I wouldn't be surprised if the PT100 RTD is more accurate than the other temp probe you are using. Temperature accuracy really isn't the big thing here it is repeatability and the controller working consistently to the temp probe, which on my machine it is perfect.

See attached for the PDF tutorial (Linux Format Magazine - what this repo was for). You can see what it looks like in there on the Rancilio Silvia. I started building another cool screen GUI (hitchhikers guide to the galaxy themed - it was amazing) for this 3 years ago then my hard drive fried itself and I lost everything. Seeing you have taken interest in this - may inspire me to start over.

[LXF216.tut_coffee.pdf](https://github.com/De-Man/silvia-pi/files/3865674/LXF216.tut_coffee.pdf)

**Comment by sjappleton** [comment]
Made some progress and updated the branch I have modified including some new icons and attempting to bind the coffee button to pin 17 to control the relay. I am still having issues getting it to shutdown and restart. So far everything I try it just freezes, even removed the need for sudo password for shutdown. Any thoughts

**Comment by sjappleton** [comment]
I have the pump and heater working now from the relays. I added some GPIO commands to the timer code.

Just going to disassemble the test unit to clean and finish the wood and metal before final assembly. Still outstanding is to test the scheduler and get the power options working.

![20191124_103137_resized](https://user-images.githubusercontent.com/26077609/69510791-d28d1000-0f03-11ea-893d-3572f013734c.jpg)
![20191124_103144_resized](https://user-images.githubusercontent.com/26077609/69510792-d28d1000-0f03-11ea-8d58-0b2a65c3242e.jpg)
![20191124_103127_resized](https://user-images.githubusercontent.com/26077609/69510793-d325a680-0f03-11ea-8cba-9c1cf74445b0.jpg)

**Comment by De-Man** [comment]
Are you able to load the background. That black background wasn’t doing the original icon sets any justice. 

Once you get the background sorted it will look much better.

**Comment by sjappleton** [comment]
I will try a background. The new buttons actually looked quite nice on the black but I might try a white background yet.

Need to get the power options working first.

**Comment by sjappleton** [comment]
Success.  I got it working now and tested the water temp control, graph/setpoints, shutdown and restart.  Tomorrow I will try making a couple shots with it. 

A couple things still need to be addressed.

1) I need to do something to ensure the PID tuning takes into account the fact that the sensor detects the boiler wall temp and not the water temp directly. As the temp quickly spikes when the heaters are since the sensor is embedded into the boiler wall and then settles to the water temp.

2) Not sure what to do with the sleep/wake functions. Looks like they were unfinished, but would it work to to change the Coffee set point to 0?  I will have to look at how to dim the screen as well.

Thanks again for all the help so far.

**Comment by De-Man** [comment]
You are most welcome.

so were you able to sort out the background in the end?

on your point (1.): this is exactly the same as the Silvia. Therefore the controller should be able to take this into consideration. You should really only need to play with the PID coefficients and the PWM duty cycle.

On your point (2.): the sleep wake functions accessed from the power menu were never finished off, all the variables are there you just need to build the logic into it. All the hard parts are done. HArd parts being: kivy date logic and formatting, it was time consuming to get it right from memory 3-4 years ago...

**Comment by sjappleton** [comment]
In the end I went without a background, just personal preference but I like the black.   I am working on the tuning but it is getting there on the tuning so far it is actually doing a decent job it is just different to look at consider some spikes you see on the graph as the al heats up quickly and then cools to the water temp.

At this time I think I am just skipping the sleep mode, currently I just use it to exit the app.

I just need one little glitch to clear up. I made the app start on boot but my graph does something funky about 10seconds where it puts in a straight line and then disappears except for a small portion. The interesting part this only happens when powering on and not restarting. I think I need the delay running the script for say 15 seconds to allow the PI to fully boot etc.

**Comment by De-Man** [comment]
Yes correct, you will need to add a 10 to 20 second delay before starting up the app. I had the same issue.

**Comment by sjappleton** [comment]
Is that just using Sleep 20 or something else?

Sean
